File: java/Easy/FindDifferenceOfTwoArrays.java
package Easy;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class FindDifferenceOfTwoArrays {
    public static void main(String[] args){
        int[] nums1 = {1,2,3,3,4,6};
        int[] nums2 = {1,1,2,2,5,5,10};
        System.out.println(findDifference2(nums1, nums2));
    }

    public static List<List<Integer>> findDifference(int[] nums1, int[] nums2){
        List<Integer> l1 = new ArrayList<Integer>();
        List<Integer> l2 = new ArrayList<Integer>();
        int[] c = new int[2001];
        for(int i: nums1) c[i + 1000] = 1;
        for(int i: nums2) c[i + 1000] = c[i+1000] == 1 || c[i+1000] == 3 ? 3 : 2; 
        for(int i: nums1) if(c[i + 1000] == 1){
            l1.add(i); c[i + 1000] = 3;
        } 
        for(int i: nums2) if(c[i + 1000] == 2){
            l2.add(i); c[i + 1000] = 3;
        } 
        return List.of(l1, l2);
    }

    public static List<List<Integer>> findDifference2(int[] nums1, int[] nums2){
        List<Integer> a1 = new ArrayList<Integer>();
        List<Integer> a2 = new ArrayList<Integer>();
        Set<Integer> s = new HashSet<>();
        Set<Integer> s2 = new HashSet<>();
        for(int i: nums1){
            s.add(i);
            s2.add(i);
        } 
        for(int i: nums2){
            if(s.contains(i)){
                s2.remove(i);
            } else {
                s.add(i);
                a2.add(i);
            }
        }
        for(int i: s2) a1.add(i);
        return List.of(a1, a2);
    }

}

File: javascript/Easy/FindDifferenceOfTwoArrays.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[][]}
 */
var findDifference = function(nums1, nums2){
    const l1 = [];
    const l2 = [];
    let c = [];
    let j = 0;
    for(let i of nums1) c[i+1000] = 1; 
    for(let i of nums2) {
        if(c[i+1000] == 1){
            c[i+1000] = 2;
        } else if(c[i+1000] != 2) {
            l2[j++] = i; c[i+1000] = 2;
        }
    }
    j = 0;
    for(let i of nums1) if(c[i+1000] == 1) {
            l1[j++] = i; c[i+1000] = 2;
        }
    return [l1, l2];
}

var findDifference2 = function(nums1, nums2){
    const s1 = new Set(nums1);
    const s2 = new Set(nums2);
    const l1 = [...s1].filter(i => !s2.has(i))
    const l2 = [...s2].filter(i => !s1.has(i))
    return [l1, l2];
}

var findDifference3 = function(nums1, nums2){
    const l1 = new Set(nums1);
    const l2 = new Set(nums2);
    for(let i of l2) if(l1.has(i)){
           l1.delete(i);
           l2.delete(i);
        }
    return [[...l1], [...l2]];
}

let a = [1, 3, 3, 7, 4, 5];
let b = [2, 4, 6, 7, 6];
console.log(findDifference3(a, b))